### ISSUE TITLE
Boolean conditional type is not narrowed correctly

### ISSUE THREAD

**Issue by alesmenzel** [microsoft/TypeScript#45272]
# Bug Report

### 🔎 Search Terms

boolean, narrowing, type guard

### 🕗 Version & Regression Information

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8UB2BXAtgIwgJwFCkigCEEoBnYbASyQHN9xoBhAHgA0A+UtqCAD2AQkAEzJRKKaAH5YUAFzFcuAMYB7JBSgBDBXEQAWAEwr1mjApKIA5Blu2rJjcChqRVYFXWkAZloA2ZBBK3ihIyh5etBDAAGr+klAAFACUCqwEEKreLurC7p5IXADeuFBQVNmJrnkRSMlQ2NEo2EjaUAD07eIAFlRiZN2qKH7CyKrOjWCq2M5aYgDuWi00tOQ0ytDAvWIY2Fph3eViSBAAbjgNEFrK3RCjpZfAza0YuAC+jpqn8dCIUbE-FIdLpqbCNcLiRhAA)

### 💻 Code

```ts
type A = number
type B = string
type C<X> = X extends true ? A : B

const a: A = 42
const b: B = 'bbbbb'
const condition = false

function getValue (): C<typeof condition> {
  if (condition) return a // this should not report as warning since this branch is never reached 
  return b
}
const value = getValue() // correct type
```

### 🙁 Actual behavior

TS outputs errors of a wrong type since it doesn't recognise that the if branch is not reachable.

### 🙂 Expected behavior

There should be no errors, the if branch will never reach.

**Comment by MartinJohns** [comment]
This is working as intended.

---

Your variable `condition` is **not** typed as `boolean`, but it is typed as `false`. As a result `C<typeof condition>` evaluates to `string`, and trying to return a `number` (line 10) results in an error.

You can fix this by either making `condition` mutable (use `let`), or add a type assertion to force a more wide type (`false as boolean`). In both cases the variable will be typed `boolean` instead of `false`, resulting in `C<typeof condition>` to be typed `string | number`. This allows your function to either return a `string` or a `number`.

---

If your issue is actually about receiving a type error in a code path that won't ever be executed, then you can drastically simplify the example:

    function getValue(): string {
      if (false) return 1;
      return "";
    }

My wager is that this is intentional as well.
